rand_core/src/impls.rs

```diff
@@ -56,41 +53,14 @@ pub fn fill_bytes_via_next<R: RngCore + ?Sized>(rng: &mut R, dest: &mut [u8]) {
     }
 }
 
-macro_rules! impl_uint_from_fill {
-    ($rng:expr, $ty:ty, $N:expr) => {{
-        debug_assert!($N == size_of::<$ty>());
-
-        let mut int: $ty = 0;
-        unsafe {
-            let ptr = &mut int as *mut $ty as *mut u8;
-            let slice = slice::from_raw_parts_mut(ptr, $N);
-            $rng.fill_bytes(slice);
-        }
-        int
-    }};
-}
-
 macro_rules! fill_via_chunks {
-    ($src:expr, $dst:expr, $ty:ty, $size:expr) => {{
-        let chunk_size_u8 = min($src.len() * $size, $dst.len());
-        let chunk_size = (chunk_size_u8 + $size - 1) / $size;
-        if cfg!(target_endian = "little") {
-            unsafe {
-                copy_nonoverlapping(
-                    $src.as_ptr() as *const u8,
-                    $dst.as_mut_ptr(),
-                    chunk_size_u8);
-            }
-        } else {
-            for (&n, chunk) in $src.iter().zip($dst.chunks_mut($size)) {
-                let tmp = n.to_le();
-                let src_ptr = &tmp as *const $ty as *const u8;
-                unsafe {
-                    copy_nonoverlapping(src_ptr,
-                                        chunk.as_mut_ptr(),
-                                        chunk.len());
-                }
-            }
+    ($src:expr, $dst:expr, $ty:ty) => {{
+        const SIZE: usize = core::mem::size_of::<$ty>();
+        let chunk_size_u8 = min($src.len() * SIZE, $dst.len());
+        let chunk_size = (chunk_size_u8 + SIZE - 1) / SIZE;
+
+        for (&n, chunk) in $src.iter().zip($dst.chunks_mut(SIZE)) {
+            chunk.copy_from_slice(&n.to_le_bytes()[..chunk.len()]);
         }
```

Comment:
~~I believe this is not equivalent to the old method: `clone_from_slice` panics if the two slices have different lengths, while the above only copies `chunk.len()` bytes.~~ Fixed.

And does this optimise down to a mem-copy?

Reply:
Unfortunately not. I have a local version with `chunks_exact_mut` that generates better code, but is quite a bit more complicated. It's still slower than the old version. I'll push it.

Reply:
The following code [optimizes](https://rust.godbolt.org/z/94hahj) down to `memcpy` (plus some code which can be removed if compiler will verify that buffer is not empty and they have equal length in bytes):
```rust
for (chunk, n) in buf.chunks_exact_mut(4).zip(val.iter()) {
    chunk.copy_from_slice(&n.to_le_bytes());
}
```

Reply:
This is similar to what I currently have. It is complicated by having to account for the case where the buffer sizes don't match, because `into_remainder` requires not moving the iterator, which rules out for loops and `zip`:

```rust
pub fn fill_via_u32_chunks(src: &[u32], dst: &mut [u8]) {
    let mut iter_src = src.iter();
    let mut iter_chunks = dst.chunks_exact_mut(4);
    let mut n;
    loop {
        n = iter_src.next();
        if let (Some(n_), Some(chunk)) = (n, iter_chunks.next()) {
            chunk.copy_from_slice(&n_.to_le_bytes())
        } else {
            break;
        }
    }
    let rem = iter_chunks.into_remainder();
    if let Some(n_) = n {
        rem.copy_from_slice(&n_.to_le_bytes()[..rem.len()]);
    }
}
```

Reply:
You can write it like this:
```
pub fn fill_via_u32_chunks(src: &[u32], dst: &mut [u8]) {
    let mut iter_src = src.iter();
    let mut chunks = dst.chunks_exact_mut(4);
    for (chunk, n) in (&mut chunks).zip(&mut iter_src) {
        chunk.copy_from_slice(&n.to_le_bytes())
    }
    let rem = chunks.into_remainder();
    if let Some(n) = iter_src.next() {
        rem.copy_from_slice(&n.to_le_bytes()[..rem.len()]);
    }
}
```

Reply:
Thanks! I changed the code accordingly.